cpp/include/kvikio/remote_handle.hpp

```diff
@@ -0,0 +1,356 @@
+      try {
+        Aws::SDKOptions options;
+        Aws::ShutdownAPI(options);
+      } catch (const std::exception& e) {
+        std::cerr << "~S3Context(): " << e.what() << std::endl;
+      }
+    }
+  }
+
+  /**
+   * @brief Get a reference to the S3 client.
+   *
+   * @return S3 client.
+   */
+  Aws::S3::S3Client& client() { return *_client; }
+
+  // No copy and move semantic
+  S3Context(S3Context const&)       = delete;
+  void operator=(S3Context const&)  = delete;
+  S3Context(S3Context const&&)      = delete;
+  void operator=(S3Context const&&) = delete;
+
+  /**
+   * @brief Get the size of a S3 file
+   *
+   * @param bucket_name The bucket name.
+   * @param object_name The object name.
+   * @return Size of the file in bytes.
+   */
+  std::size_t get_file_size(const std::string& bucket_name, const std::string& object_name)
+  {
+    KVIKIO_NVTX_FUNC_RANGE();
+    Aws::S3::Model::HeadObjectRequest req;
+    req.SetBucket(bucket_name.c_str());
+    req.SetKey(object_name.c_str());
+    Aws::S3::Model::HeadObjectOutcome outcome = client().HeadObject(req);
+    if (!outcome.IsSuccess()) {
+      const Aws::S3::S3Error& err = outcome.GetError();
+      throw std::invalid_argument("get_file_size(): " + err.GetExceptionName() + ": " +
+                                  err.GetMessage());
+    }
+    return outcome.GetResult().GetContentLength();
+  }
+};
+
+/**
+ * @brief Handle of remote file (currently, only AWS S3 is supported).
+ */
+class RemoteHandle {
+ private:
+  std::string _bucket_name{};
+  std::string _object_name{};
+  std::size_t _nbytes{};
+  std::shared_ptr<S3Context> _context;
+
+ public:
+  /**
+   * @brief Construct from a bucket and object name pair.
+   *
+   * @param context The S3 context used for the connection to the remove server.
+   * @param bucket_and_object_name Name pair <bucket, object>.
+   */
+  RemoteHandle(std::shared_ptr<S3Context> context,
+               std::pair<std::string, std::string> bucket_and_object_name)
+  {
+    if (!context) { throw std::invalid_argument("RemoteHandle(): context cannot be null"); }
+    _context     = std::move(context);
+    _bucket_name = std::move(bucket_and_object_name.first);
+    _object_name = std::move(bucket_and_object_name.second);
+    _nbytes      = _context->get_file_size(_bucket_name, _object_name);
+  }
+
+  /**
+   * @brief Construct from a bucket and object name.
+   *
+   * @param context The S3 context used for the connection to the remove server.
+   * @param bucket_name Name of the bucket.
+   * @param object_name Name of the object.
+   */
+  RemoteHandle(std::shared_ptr<S3Context> context, std::string bucket_name, std::string object_name)
+    : RemoteHandle(std::move(context),
+                   std::make_pair(std::move(bucket_name), std::move(object_name)))
+  {
+  }
+
+  /**
+   * @brief Construct from a remote path such as "s3://<bucket>/<object>".
+   *
+   * @param context The S3 context used for the connection to the remove server.
```

Comment:
```suggestion
   * @param context The S3 context used for the connection to the remote server.
```

Reply:
Heh, you don't like the _remove_ server?